### PR TITLE
Feat: support load hdr to cubemap in RGBM format and invert x axis

### PR DESCRIPTION
**Pull request by zhuxudong**
### What is the current behavior? (You can also link to an open issue here)
HDRLoader is in baker package and use RGBE format.

### What is the new behavior (if this is a feature change)?
move HDR loader to here,and use RGBM format to avoid filter error, and invert x axis to avoid reflection error. #558

### CODE REVIEW

File: packages/loader/src/HDRLoader.ts [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+      for (let x = 0; x < texSize; x++) {
+        const v = this._temp5Vector3.setValue(0, 0, 0).add(xv2).subtract(xv1).scale(fy).add(xv1);
+        v.normalize();
+
+        const color = this._calcProjectionSpherical(v, pixels, inputWidth, inputHeight);
+        this._RGBEToLinear(color);
+        this._linearToRGBM(color, 5);
+
+        // 4 channels per pixels
+        textureArray[y * texSize * 4 + x * 4] = color.r;
```
Comment (GuoLei1990) [inline packages/loader/src/HDRLoader.ts:137]: cache 'y * texSize * 4 + x * 4'

File: packages/loader/src/HDRLoader.ts [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+      py = 0;
+    } else if (py >= inputHeight) {
+      py = inputHeight - 1;
+    }
+
+    const inputY = inputHeight - py - 1;
+    const r = pixels[inputY * inputWidth * 4 + px * 4];
+    const g = pixels[inputY * inputWidth * 4 + px * 4 + 1];
+    const b = pixels[inputY * inputWidth * 4 + px * 4 + 2];
+    const a = pixels[inputY * inputWidth * 4 + px * 4 + 3];
```
Comment (GuoLei1990) [inline packages/loader/src/HDRLoader.ts:192]: cache `inputY * inputWidth * 4 + px *`

File: packages/loader/src/HDRLoader.ts [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    color.g *= scaleFactor;
+    color.b *= scaleFactor;
+    color.a = 1;
+  }
+
+  private static _linearToRGBM(color: Color, maxRange: number): void {
+    const maxRGB = Math.max(color.r, Math.max(color.g, color.b));
+    let M = Math.min(maxRGB / maxRange, 1);
+    M = Math.ceil(M * 255);
+    const scaleFactor = 65025 / (M * maxRange);
```
Comment (GuoLei1990) [inline packages/loader/src/HDRLoader.ts:356]: What is the meaning of `65025 `?

File: packages/loader/src/HDRLoader.ts [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+      for (let i = 0; i < l; i++) {
+        let off = 0;
+        data_rgba[offset] = scanline_buffer[i + off];
+        off += scanline_width; //1;
+
+        data_rgba[offset + 1] = scanline_buffer[i + off];
+        off += scanline_width; //1;
+
+        data_rgba[offset + 2] = scanline_buffer[i + off];
+        off += scanline_width; //1;
```
Comment (GuoLei1990) [inline packages/loader/src/HDRLoader.ts:332]: `//1;` is what?

File: packages/loader/src/HDRLoader.ts [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    const a = pixels[inputY * inputWidth * 4 + px * 4 + 3];
+
+    return new Color(r, g, b, a);
+  }
+
+  private static _readStringLine(uint8array: Uint8Array, startIndex: number): string {
+    let line = "";
+    let character = "";
+
+    for (let i = startIndex; i < uint8array.length - startIndex; i++) {
```
Comment (singlecoder) [inline packages/loader/src/HDRLoader.ts:201]: uint8array.length can opt

File: packages/loader/src/HDRLoader.ts [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  private static _calcProjectionSpherical(
+    vDir: Vector3,
+    pixels: Uint8Array,
+    inputWidth: number,
+    inputHeight: number
+  ): Color {
+    let theta = Math.atan2(vDir.z, -vDir.x);
+    let phi = Math.acos(vDir.y);
+
+    while (theta < -Math.PI) {
```
Comment (singlecoder) [inline packages/loader/src/HDRLoader.ts:161]: -Math.PI can pulled out alone

File: packages/loader/src/HDRLoader.ts [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    new Vector3(-1.0, 1.0, -1.0),
+    new Vector3(1.0, 1.0, -1.0)
+  ];
+
+  private static _faceBack = [
+    new Vector3(1.0, -1.0, 1.0),
+    new Vector3(-1.0, -1.0, 1.0),
+    new Vector3(1.0, 1.0, 1.0),
+    new Vector3(-1.0, 1.0, 1.0)
+  ];
```
Comment (GuoLei1990) [inline packages/loader/src/HDRLoader.ts:70]: Write like this is better, less memory, less code:
```
 private static _rightBottom = new Vector3(1.0, -1.0, -1.0);
  private static _rightUp = new Vector3(1.0, -1.0, -1.0);
  private static _rightFront = new Vector3(1.0, -1.0, -1.0);
  private static _rightBack = new Vector3(1.0, -1.0, -1.0);
  private static _leftBottom = new Vector3(1.0, -1.0, -1.0);
  private static _leftUp = new Vector3(1.0, -1.0, -1.0);
  private static _leftFront = new Vector3(1.0, -1.0, -1.0);
  private static _leftBack = new Vector3(1.0, -1.0, -1.0);

  private static _faceRight = [HDRLoader._rightBottom, HDRLoader._rightUp, HDRLoader._leftUp, HDRLoader._leftBottom];
  private static _faceLeft = [HDRLoader._rightBottom, HDRLoader._rightUp, HDRLoader._leftUp, HDRLoader._leftBottom];
  private static _faceUp = [HDRLoader._rightBottom, HDRLoader._rightUp, HDRLoader._leftUp, HDRLoader._leftBottom];
  private static _faceBottom = [HDRLoader._rightBottom, HDRLoader._rightUp, HDRLoader._leftUp, HDRLoader._leftBottom];
  private static _faceFront = [HDRLoader._rightBottom, HDRLoader._rightUp, HDRLoader._leftUp, HDRLoader._leftBottom];
  private static _faceBack = [HDRLoader._rightBottom, HDRLoader._rightUp, HDRLoader._leftUp, HDRLoader._leftBottom];
```

File: packages/loader/src/HDRLoader.ts [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+    inputWidth: number,
+    inputHeight: number
+  ): Color {
+    let theta = Math.atan2(vDir.z, -vDir.x);
+    let phi = Math.acos(vDir.y);
+
+    while (theta < -Math.PI) {
+      theta += 2 * Math.PI;
+    }
+    while (theta > Math.PI) {
```
Comment (singlecoder) [inline packages/loader/src/HDRLoader.ts:164]: Math.PI can pulled out alone

File: packages/loader/src/HDRLoader.ts [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+      }
+
+      num_scanlines--;
+    }
+
+    return data_rgba;
+  }
+
+  private static _RGBEToLinear(color: Color): void {
+    const scaleFactor = Math.pow(2, color.a - 136);
```
Comment (GuoLei1990) [inline packages/loader/src/HDRLoader.ts:345]: what is the meaning of `136 `

File: packages/loader/src/HDRLoader.ts [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        throw "HDR Bad header format, wrong scan line width";
+      }
+
+      // read each of the four channels for the scanline into the buffer
+      // first red, then green, then blue, then exponent
+
+      let ptr = 0,
+        count;
+
+      while (ptr < ptr_end && pos < buffer.byteLength) {
```
Comment (singlecoder) [inline packages/loader/src/HDRLoader.ts:296]: buffer.byteLength can pulled out

File: packages/loader/src/HDRLoader.ts (hunk 7 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:70]: done

File: packages/loader/src/HDRLoader.ts (hunk 1 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:137]: done

File: packages/loader/src/HDRLoader.ts (hunk 8 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:164]: done

File: packages/loader/src/HDRLoader.ts (hunk 6 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:161]: done

File: packages/loader/src/HDRLoader.ts (hunk 2 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:192]: done

File: packages/loader/src/HDRLoader.ts (hunk 5 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:201]: won't fix,length is dynamic

File: packages/loader/src/HDRLoader.ts (hunk 10 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:296]: done

File: packages/loader/src/HDRLoader.ts (hunk 4 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:332]: done

File: packages/loader/src/HDRLoader.ts (hunk 9 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:345]: done

File: packages/loader/src/HDRLoader.ts (hunk 3 above)
Comment (zhuxudong) [reply packages/loader/src/HDRLoader.ts:356]: done

**Pull request merged**